packages/aws-cdk-lib/assertions/test/template.test.ts

```diff
@@ -1345,6 +1345,29 @@ describe('Template', () => {
     }).toThrow(/dependency cycle/);
   });
 
+  test('throws when given a more complex template with cyclic dependencies', () => {
```

Comment:
While i can see how the change in `cyclic.ts` makes the `recurse` function more efficient, this test isn't going to show that one way or another right? the test will succeed with the old code as well?

I'm prepared to just approve as is anyway, because this looks low risk and is hard to test, but want to make sure i have the whole picture first.

Reply:
I initially ran the test with the old code (unchanged) and it fails for me because of a stackoverflow error. 
<img width="570" alt="Screenshot 2023-08-17 at 5 46 06 PM" src="https://github.com/aws/aws-cdk/assets/5008987/5303fb2c-99db-48b4-8df9-948e4c766b6f">

I think the call stack in the test case I added ends up looking like this (without the fix applied):

```
findCycle(deps)
recurse("Res1", ["Res1"])
recurse("Res2", ["Res1", "Res2"])
recurse("Res3", ["Res1", "Res2", "Res3"])
recurse("Res2", ["Res1", "Res2", "Res3", "Res2"])
recurse("Res3", ["Res1", "Res2", "Res3", "Res2", "Res3"])
...
```

Reply:
Ah i see